Makefile

```diff
@@ -179,22 +154,104 @@ install-deps:
 deploy:
 	./scripts/deploy.staging.sh satellite storjlabs/satellite:${TAG}
 	for i in $(shell seq 1 60); do \
-		./scripts/deploy.staging.sh storage-node-$$i storjlabs/storage-node:${TAG}; \
+		./scripts/deploy.staging.sh storagenode-$$i storjlabs/storagenode:${TAG}; \
 	done
 
-release-osx:
-	GOOS=darwin GOARCH=amd64 go build -o release/uplink-osx-amd64/uplink ./cmd/uplink
-	cd release; tar czvf uplink-osx-amd64.tar.gz uplink-osx-amd64
-	rm -rf release/uplink-osx-amd64
+.PHONY: binary
+binary: CUSTOMTAG = -${GOOS}-${GOARCH}
+binary:
+	mkdir -p release/${TAG}
+	CUSTOMTAG=$(CUSTOMTAG) $(MAKE) $(COMPONENT)-image
+	cid=$$(docker create storjlabs/$(COMPONENT):${TAG}${CUSTOMTAG}) \
+	&& docker cp $$cid:/app/$(COMPONENT) release/${TAG}/$(COMPONENT)_${GOOS}_${GOARCH} \
+    && docker rm $$cid
+	docker rmi storjlabs/$(COMPONENT):${TAG}${CUSTOMTAG}
+	rm -f release/${TAG}/${COMPONENT}_${GOOS}_${GOARCH}.zip
+	cd release/${TAG}; zip ${COMPONENT}_${GOOS}_${GOARCH}.zip ${COMPONENT}_${GOOS}_${GOARCH}
+	rm -f release/${TAG}/${COMPONENT}_${GOOS}_${GOARCH}
+
+# To update this section, modify and run the following:
+# for c in satellite storagenode uplink; do \
+# for oa in "darwin amd64" "freebsd 386" "freebsd amd64" "linux 386" \
+# "linux amd64" "windows 386" "windows amd64"; do \
+# echo "$c $oa"; done; done | while read -r c o a; do; \
+# printf ".PHONY: ${c}_${o}_${a}\n${c}_${o}_${a}:\n\tGOOS=${o} GOARCH=${a} COMPONENT=${c} \$(MAKE) binary\n"; \
+# done
+.PHONY: satellite_darwin_amd64
+satellite_darwin_amd64:
```

Comment:
Any reason not to use:

```
satellite_%_amd64:
	GOOS=$@ GOARCH=amd64 COMPONENT=satellite $(MAKE) binary
satellite_%_386:
	GOOS=$@ GOARCH=amd386 COMPONENT=satellite $(MAKE) binary
```

or (using echo to make it more obvious

```
.PHONY: all

COMPONENTLIST := uplink satellite
OSLIST        := windows linux darwin
ARCHLIST      := amd64 386

BINARIES := $(foreach C,$(COMPONENTLIST),$(foreach O,$(OSLIST),$(foreach A,$(ARCHLIST),build_$C_$O_$A)))

all: $(BINARIES)

build_%:
	@echo COMPONENT=$(word 1, $(subst _, ,$@))
	@echo GOOS=$(word 2, $(subst _, ,$@))
	@echo GOARCH=$(word 3, $(subst _, ,$@))
	@echo COMPONENT=$(word 1, $(subst _, ,$@)) GOOS=$(word 2, $(subst _, ,$@)) GOARCH=$(word 3, $(subst _, ,$@)) $(MAKE) binary
```

Reply:
We're not building everything everywhere, only linux arm, darwin only amd64.

Reply:
We could try to build everything everywhere, then fail if some combinations don't exist, but that feels weird too.

Reply:
Note, you can also manually specify the BINARIES, or have multiple foreaches. e.g.

```
COMPONENTLIST := uplink satellite
OSARCHLIST    := windows_amd64 windows_386 linux_arm darwin_amd64

BINARIES := $(foreach C,$(COMPONENTLIST),$(foreach O,$(OSARCHLIST),build_$C_$O))
```